core/src/main/java/org/infinispan/interceptors/distribution/L1LastChanceInterceptor.java

```diff
@@ -76,40 +76,50 @@ public void start() {
    }
 
    public CompletableFuture<Void> visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command, boolean assumeOriginKeptEntryInL1) throws Throwable {
-      Object returnValue = ctx.forkInvocationSync(command);
-      Object key;
-      if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
-            cdl.localNodeIsOwner((key = command.getKey()))) {
-         if (trace) {
-            log.trace("Sending additional invalidation for requestors if necessary.");
+      ctx.onReturn((ctx1, command1, rv, throwable) -> {
+         if (throwable != null)
+            throw throwable;
+
+         Object key;
+         if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
```

Comment:
On Wed, May 18, 2016 at 2:05 PM, Pedro Ruivo <notifications@github.com>
wrote:

> In
> core/src/main/java/org/infinispan/interceptors/distribution/L1LastChanceInterceptor.java
> <https://github.com/infinispan/infinispan/pull/4324#discussion_r63683816>:
>
> > @@ -76,40 +76,50 @@ public void start() {
> >     }
> >
> >     public CompletableFuture<Void> visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command, boolean assumeOriginKeptEntryInL1) throws Throwable {
> > -      Object returnValue = ctx.forkInvocationSync(command);
> > -      Object key;
> > -      if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
> > -            cdl.localNodeIsOwner((key = command.getKey()))) {
> > -         if (trace) {
> > -            log.trace("Sending additional invalidation for requestors if necessary.");
> > +      ctx.onReturn((ctx1, command1, rv, throwable) -> {
> > +         if (throwable != null)
> > +            throw throwable;
> > +
> > +         Object key;
> > +         if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
>
> I have a couple of suggestions:
>
>    - change the arguments name. ctx1 is kind of weird to read. you can
>    use a r before each argument: rCtx, rCommand, rValue, rThrowable to
>    show that is the return.
>    - creating fields to store the ReturnHandler is also strange. wouldn't
>    be better to create a private method (example CompletableFuture<Object>
>    doSomethingOnReturn(...)) and do
>    ctx.onReturn(this::doSomethingOnReturn).
>
>
I guess I could go with `rCtx` and `rCommand`, maybe even `rValue`, but
`rThrowable` just looks odd :)